File: binary_search_tree.py
# A node in a Binary Tree has data, left, and right pointers
# The left and right pointers are intialized as null
class BinaryNode:
    def __init__(self, data):
      self.left = None
      self.right = None
      self.data = data


class BinaryTree:
    # When a new Tree is initialized, it has a null root property
    # This means the Tree begins as empty
    def __init__(self):
        self.root = None

    # Insert creates a new Node from the data passed in and adds it to the tree
    # If the data is already in the tree, do not insert it again
    def insert(self, data):
        new_node = BinaryNode(data)
        if not self.root:
            self.root = new_node
            return

        walker = self.root
        while walker:
            if data < walker.data:
                if not walker.left:
                    walker.left = new_node
                    return
                else: walker = walker.left
            elif data > walker.data:
                if not walker.right:
                    walker.right = new_node
                    return
                else: walker = walker.right

    # Search the Tree for a node with the given value
    # If the node exists, return it
    # If the node doesn't exist, return false
    def search(self, val):
        if not self.root:
            return False

        walker = self.root
        while walker:
            if val < walker.data:
                walker = walker.left
            elif val > walker.data:
                walker = walker.right
            else: return walker

        return False #  Outside loop means we did not find it!

    # Calculate the number of nodes in the tree, starting from the given node
    # If no node is provided, we can use the root as a sensible default
    # def size(self, node=None):
    #     if not node:
    #         node = self.root

    #     count = 0
    #     def recursive_size(node):
    #         if node:
    #             count += 1
    #             recursive_size(node.left)
    #             recursive_size(node.right)
    #     recursive_size(node)
    #     return count

    # Return the maximum data value stored in the tree
    def getMax(self):
        if not self.root:
            return None

        walker = self.root
        while walker.right:
            walker = walker.right

        return walker.data

    # Calculate the maximum amount of nodes in any one path from the given node
    # If not given a specific node, default to using the root node
    # def height(self, node=None):
    #     if not node:
    #         node = self.root

    #     max_height = 0
    #     def recursive_height(node, height=1):
    #         global max_height
    #         print(max_height)
    #         if node:
    #             if height > max_height:
    #                 max_height = height

    #             recursive_height(node.left, height + 1)
    #             recursive_height(node.right, height + 1)


    #     recursive_height(node)
    #     return max_height

# Some Test Code
my_bst = BinaryTree()
my_bst.insert(5)
my_bst.insert(2)
my_bst.insert(1)
my_bst.insert(3)
my_bst.insert(10)
my_bst.insert(7)
my_bst.insert(6)
my_bst.insert(9)
my_bst.insert(8)
my_bst.insert(4)
print('Height', my_bst.height())
print('Size', my_bst.size())
print('Max Val', my_bst.getMax())

File: binarySearchTree.js
// A node in a Binary Tree has data, left, and right pointers
// The left and right pointers are intialized as null
class BinaryNode {
  constructor(data){
    this.data = data;
    this.left = null;
    this.right = null;
  }
}

class BinaryTree {
  // When a new Tree is initialized, it has a null root property
  // This means the Tree begins as empty
  constructor() {
    this.root = null;
  }

  // Insert creates a new Node from the data passed in and adds it to the tree
  // If the data is already in the tree, do not insert it again
  insert(data) {
    const newNode = new BinaryNode(data);
    if (!this.root) {
        this.root = newNode;
        return this;
    }

    let walker = this.root;
    while (walker){
      if (data < walker.data) {
        if (!walker.left) {
          walker.left = newNode;
          return this;
        }
        else {
          walker = walker.left
        }
      }
      else if (data > walker.data) {
        if (!walker.right) {
          walker.right = newNode;
          return this;
        }
        else {
          walker = walker.right;
        }
      }
    }
  }

  // Search the Tree for a node with the given value
  // if the node exists, return it
  // if the node doesn't exist, return false
  search(val) {
    if (!this.root) {
      return false;
    }

    let walker = this.root;
    while (walker) {
      if (val < walker.data) {
        walker = walker.left;
      }
      else if(val > walker.data) {
        walker = walker.right;
      }
      else {
        return walker;
      }
    }

    return false;
  }

  // Calculate the number of nodes in the tree, starting from the given node
  // If no node is provided, we can use the root as a sensible default
  size(node=this.root) {
    let count = 0;
    function rSize(node) {
      if (node) {
        count++;
        rSize(node.left);
        rSize(node.right);
      }
    }
    rSize(node);
    return count;
  }

  // Return the maximum data value stored in the tree
  getMax() {
    if (!this.root) {
      return null;
    }
    let walker = this.root;
    while (walker.right) {
      walker = walker.right;
    }
    return walker.data;
  }

  // Calculate the maximum amount of nodes in any one path from the given node
  // If not given a specific node, default to using the root node
  height(node=this.root) {
    let maxHeight = 0;
    function rHeight(node, height=1) {
      if (node) {
        if (height > maxHeight) {
          maxHeight = height;
        }
        rHeight(node.left, height+1);
        rHeight(node.right, height+1);
      }
    }
    rHeight(node);
    return maxHeight;
  }
}

// Some Test Code
let myBst = new BinaryTree();
myBst.insert(5);
myBst.insert(2);
myBst.insert(1);
myBst.insert(3);
myBst.insert(10);
myBst.insert(7);
myBst.insert(6);
myBst.insert(9);
myBst.insert(8);
myBst.insert(4);
console.log('Height', myBst.height());
console.log('Size', myBst.size());
console.log('Max Val', myBst.getMax());
